File: jeffrey-microscope/pages-microscope/src/views/global/advisorParallelism.test.ts
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { describe, expect, it } from 'vitest';
import {
  DEFAULT_MAX_CONCURRENT_RUNS,
  keyForValue,
  PARALLELISM_CHOICES,
  UNLIMITED
} from '@/views/global/advisorParallelism';

describe('keyForValue', () => {
  it('selects the named card for each value it stores', () => {
    expect(keyForValue('1')).toBe('one');
    expect(keyForValue(String(DEFAULT_MAX_CONCURRENT_RUNS))).toBe('balanced');
    expect(keyForValue(String(UNLIMITED))).toBe('unlimited');
  });

  // A value nobody can reach by clicking must still reopen showing its real number, rather than being
  // rounded onto the nearest card.
  it('treats anything else as Custom', () => {
    expect(keyForValue('3')).toBe('custom');
    expect(keyForValue('16')).toBe('custom');
  });

  it('falls back to the default when nothing is stored or the value is not a number', () => {
    expect(keyForValue(undefined)).toBe('balanced');
    expect(keyForValue(null)).toBe('balanced');
    expect(keyForValue('')).toBe('balanced');
    expect(keyForValue('not-a-number')).toBe('balanced');
  });

  it('accepts a number as readily as its string form', () => {
    expect(keyForValue(1)).toBe('one');
    expect(keyForValue(0)).toBe('unlimited');
    expect(keyForValue(7)).toBe('custom');
  });
});

describe('PARALLELISM_CHOICES', () => {
  it('offers exactly one card per key, in the order the page renders them', () => {
    expect(PARALLELISM_CHOICES.map(c => c.key)).toEqual(['one', 'balanced', 'unlimited', 'custom']);
  });

  // Every card that stores a value must round-trip back to itself, or clicking one would select another.
  it('round-trips each stored value back to its own card', () => {
    for (const choice of PARALLELISM_CHOICES) {
      if (choice.value !== null) {
        expect(keyForValue(choice.value)).toBe(choice.key);
      }
    }
  });

  it('leaves Custom without a value of its own, since it keeps whatever the field holds', () => {
    expect(PARALLELISM_CHOICES.find(c => c.key === 'custom')?.value).toBeNull();
  });

  // The number of event types varies per recording, so no card may promise a count.
  it('names no event-type count on the unlimited card', () => {
    const unlimited = PARALLELISM_CHOICES.find(c => c.key === 'unlimited');
    expect(unlimited?.badge).toBe('All');
    expect(`${unlimited?.description} ${unlimited?.estimate}`).not.toMatch(/\d/);
  });
});

File: jeffrey-microscope/profiles/profile-advisor/src/test/java/cafe/jeffrey/profile/advisor/settings/AdvisorParallelismTest.java
/*
 * Jeffrey
 * Copyright (C) 2026 Petr Bouda
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cafe.jeffrey.profile.advisor.settings;

import cafe.jeffrey.profile.common.pipeline.PipelineRunOptions;
import cafe.jeffrey.shared.common.config.MicroscopeSettingKeys;
import cafe.jeffrey.shared.common.config.SettingsStore;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

class AdvisorParallelismTest {

    /**
     * The key has to be declared for a stored value to be readable at all — an undeclared name is
     * dropped as unknown, which is how the real store keeps a stale row from resurrecting a setting.
     */
    private static AdvisorParallelism withStored(String value) {
        SettingsStore store = new SettingsStore(
                Map.of(MicroscopeSettingKeys.ADVISOR_MAX_CONCURRENT_RUNS, "2"),
                Map.of(MicroscopeSettingKeys.ADVISOR_MAX_CONCURRENT_RUNS, value));
        return new AdvisorParallelism(store);
    }

    @Nested
    class StoredCeiling {

        @Test
        void isUsedAsWritten() {
            assertEquals(3, withStored("3").getAsInt());
        }

        @Test
        void ofOneIsAllowed() {
            assertEquals(1, withStored("1").getAsInt());
        }
    }

    @Nested
    class Unlimited {

        // The settings page cannot offer "every type" as a number — a recording is analyzed only for the
        // types it carries — so zero is what All stores.
        @Test
        void isWhatZeroMeans() {
            assertEquals(PipelineRunOptions.UNBOUNDED, withStored("0").getAsInt());
        }
    }

    @Nested
    class UnusableValue {

        @Test
        void fallsBackToTheDefaultRatherThanFailingTheRun() {
            assertEquals(AdvisorParallelism.DEFAULT_MAX_CONCURRENT_RUNS, withStored("-1").getAsInt());
            assertEquals(AdvisorParallelism.DEFAULT_MAX_CONCURRENT_RUNS, withStored("not-a-number").getAsInt());
        }

        @Test
        void undeclaredFallsBackToTheDefault() {
            SettingsStore empty = new SettingsStore(Map.of(), Map.of());
            assertEquals(
                    AdvisorParallelism.DEFAULT_MAX_CONCURRENT_RUNS,
                    new AdvisorParallelism(empty).getAsInt());
        }
    }
}
